randovania/game_description/resources/resource_collection.py

```diff
@@ -0,0 +1,153 @@
+        return f"<ResourceCollection with {self.num_resources} resources>"
+
+    @property
+    def _comparison_tuple(self):
+        return list(self.as_resource_gain()), self.add_self_as_requirement_to_resources
+
+    def __eq__(self, other):
+        return isinstance(other, ResourceCollection) and (
+                self._comparison_tuple == other._comparison_tuple
+        )
+
+    @property
+    def num_resources(self):
+        return len(self._existing_resources)
+
+    def has_resource(self, resource: ResourceInfo) -> bool:
+        return self[resource] > 0
+
+    def is_resource_set(self, resource: ResourceInfo) -> bool:
+        """
+        Checks if the given resource has a value explicitly set, instead of using the fallback of 0.
+        :param resource:
+        :return:
+        """
+        return resource.resource_index in self._existing_resources
+
+    def set_resource(self, resource: ResourceInfo, quantity: int):
+        resource_index = resource.resource_index
+        self._damage_reduction_cache = None
+        try:
+            self._resource_array[resource_index] = quantity
+        except IndexError:
+            self._resize_array_to(resource_index)
+            self._resource_array[resource_index] = quantity
+        self._existing_resources[resource_index] = resource
+
+        mask = 1 << resource_index
+        if quantity > 0:
+            self.resource_bitmask |= mask
+        elif self.resource_bitmask & mask:
+            self.resource_bitmask -= mask
+
+    @classmethod
+    def from_dict(cls, db: ResourceDatabase, resources: dict[ResourceInfo, int]) -> ResourceCollection:
+        result = cls.with_database(db)
+        result.add_resource_gain(resources.items())
+        return result
+
+    @classmethod
+    def from_resource_gain(cls, db: ResourceDatabase, resource_gain: ResourceGain) -> ResourceCollection:
+        result = cls.with_database(db)
+        result.add_resource_gain(resource_gain)
+        return result
+
+    def add_resource_gain(self, resource_gain: ResourceGain):
+        self._damage_reduction_cache = None
+        for resource, quantity in resource_gain:
+            resource_index = resource.resource_index
+            try:
+                self._resource_array[resource_index] += quantity
+            except IndexError:
+                self._resize_array_to(resource_index)
+                self._resource_array[resource_index] += quantity
+            self._existing_resources[resource_index] = resource
+
+            mask = 1 << resource_index
+            if self._resource_array[resource_index] > 0:
+                self.resource_bitmask |= mask
+            elif self.resource_bitmask & mask:
+                self.resource_bitmask -= mask
+
+    def as_resource_gain(self) -> ResourceGain:
+        for index, resource in self._existing_resources.items():
+            yield resource, self._resource_array[index]
+
+    def remove_resource(self, resource: ResourceInfo):
```

Comment:
and `__delitem__`?

Reply:
Doing `del collection[my_item]` is such a weird API

Reply:
yeah I can live without that one lmao I don't think I've ever used `del` in python a single time